### DISCUSSION TITLE
Spectrum flash of unstyled content (FOUC)

### DISCUSSION THREAD

**Discussion by cheema-corellian** [adobe/react-spectrum#6462] [Q&A]
Has anybody noticed a flash of unstyled content ([FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content)) with Spectrum React on page load?  At first I thought I was doing something wrong, but then I was able to reproduce the issue in a brand new sandbox repo with just the Spectrum React button component and nothing else.  With and without the Adobe Clean font.  And then I noticed at least one person on YouTube talk about it as being a problem as well.

I can create a Stackblitz example if it is needed.

Has anybody else around here noticed it?  Is there a solution or workaround for it?

In Next.js, it is very visible in dev builds and a little more subtle in production builds.

**Comment by snowystinger** [comment]
I'm not sure what we can do about it, we rely on media queries to determine light or dark mode. Those don't work until you're on the client. You could set the colorScheme on your provider on the server if you know ahead of time through a user setting, then you shouldn't have the issue. https://react-spectrum.adobe.com/react-spectrum/Provider.html#color-schemes

If I've missed the gist here, a stackblitz would definitely be helpful

**Comment by nicksergeant** [comment]
I've been noticing this as well, and it is pretty jarring -- both the flash of mobile scale on desktop, and of light mode on dark mode:

https://github.com/user-attachments/assets/a6bb26f0-ac7e-4be4-8a47-905c92240f5e

Here is a very simple POC demonstrating the issue:

- Preview: https://nextjs-spectrum.vercel.app/
- Repo: https://github.com/nicksergeant/nextjs-spectrum/

I've been trying to figure out how to get around this issue for a while now and I'm stumped. I'm at the point now where I'm about to dive into the source of https://react-spectrum.adobe.com/react-spectrum/ itself because for some reason, the issue does not present at all:

https://github.com/user-attachments/assets/9e5956bb-db21-425a-8882-186c2f3c42e4

**Reply by obsicn** [reply]
@nicksergeant Have you solved this problem? I faced the same problem when using react-spectrum with Next.JS 15. Thanks. Michael

**Reply by nicksergeant** [reply]
@obsicn nope, I sadly gave up trying to use react-spectrum w/ Next.js as this is a pretty janky experience that I couldn't figure out how to get around.

**Reply by snowystinger** [reply]
We store the theme in localStorage for our docs, I'd bet you'd still see the flash if you hard reloaded ours. If you're ok with it only happening on initial load, this should work for you as well. You might need to suppress a hydration warning, but that's the only solution I know of unless you can get the information to the server to render it correctly initially as well

Note, this shouldn't be an issue in our implementation of the upcoming Spectrum 2 because we were able to take advantage of newer browser features. https://react-spectrum.adobe.com/s2/index.html

**Reply by obsicn** [reply]
Good news. Are there any  tutorial and sample  about how to use React Spectrum  (including Spectrum 2) with Next.js 15? Which build tool is recommended when using React Spectrum on production? Parcel, Webpack, Next.js, Vite, Rollup, or ESBuild? Thanks.

**Reply by snowystinger** [reply]
We have some example setups here https://github.com/adobe/react-spectrum/tree/main/examples

We don't have much documentation specific to Next.js.
These are the getting started docs though for general use.
https://react-spectrum.adobe.com/react-spectrum/getting-started.html
https://react-spectrum.adobe.com/s2/index.html?path=/docs/style-macro--docs

And you can search the docs for Next.js references, such https://react-spectrum.adobe.com/react-spectrum/ssr.html#nextjs

**Reply by obsicn** [reply]
@snowystinger Thank. I have tried React Spectrum 2 and the flash problem has been solved according to the sample code https://github.com/adobe/react-spectrum/tree/main/examples/s2-next-macros. When will React Spectrum 2 formal release be available?

**Reply by snowystinger** [reply]
It's in pre-release right now as we work towards "feature parity"*^tm, which we don't have a date for yet. That said, the components we do have are fairly stable, they're built on the stable React Aria Components and mostly apply styles or add virtualisation. We make the occasional API change for them, which is why it's pre-release at the moment, but we should add anything of note to our release notes.
